src/ControlSystem/TimescaleTuner.cpp

```diff
@@ -0,0 +1,126 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ControlSystem/TimescaleTuner.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <ostream>
+
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/Gsl.hpp"
+
+TimescaleTuner::TimescaleTuner(DataVector initial_timescale,
+                               double max_timescale, double min_timescale,
+                               double max_threshold, double min_threshold,
+                               double increase_factor,
+                               double decrease_factor) noexcept
+    : timescale_{std::move(initial_timescale)},
+      max_timescale_{max_timescale},
+      min_timescale_{min_timescale},
+      max_threshold_{max_threshold},
+      min_threshold_{min_threshold},
+      increase_factor_{increase_factor},
+      decrease_factor_{decrease_factor} {
+  for (auto t_scale : timescale_) {
+    if (t_scale <= 0.0) {
+      ERROR("Initial timescale must be > 0");
+    }
+  }
+
+  if (decrease_factor_ > 1.0 or decrease_factor <= 0.0) {
+    ERROR("The specified decrease_factor "
+          << decrease_factor_ << " must satisfy 0 < decrease_factor <= 1");
+  }
+  if (increase_factor_ < 1.0) {
+    ERROR("The specified increase factor " << increase_factor_
+                                           << " must be >= 1.0");
+  }
+  if (min_timescale_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_timescale_ <= min_timescale_) {
+    ERROR("The maximum timescale "
+          << max_timescale_
+          << " must be > than the specified minimum timescale "
+          << min_timescale_);
+  }
+  if (min_threshold_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_threshold_ <= min_threshold_) {
+    ERROR("The maximum threshold "
+          << max_threshold_
+          << " must be > than the specified minimum threshold "
+          << min_threshold_);
+  }
+}
+
+void TimescaleTuner::suggested_timescale_if_in_allowable_range(
+    double suggested_timescale) noexcept {
+  for (auto& t_scale : timescale_) {
+    if (suggested_timescale > max_timescale_) {
+      t_scale = max_timescale_;
+    } else if (suggested_timescale < min_timescale_) {
+      t_scale = min_timescale_;
+    } else {
+      t_scale = suggested_timescale;
+    }
+  }
+}
+
+void TimescaleTuner::update_timescale(
+    const std::array<DataVector, 3>& qs) noexcept {
+  // above, Q is an array of length 3: \int Q, Q, dQ/dt
+  // each of these 3 is a DataVector w/ length = the number of components
+
```

Comment:
Are we hardcoding PID in this function? I prefer that we do, but if we don't, then this comment should be generalized to the PND case.

Reply:
I've cleaned this up. We don't actually need the integral term in the TimescaleTuner, so I removed that. Also, the TimescaleTuner is designed to be independent of the controller, so this should work for PID or PND.